### ISSUE TITLE
can not reproduce the result in log correctly

### ISSUE THREAD

**Issue by CrazyBrick** [zcablii/LSKNet#66]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmrotate/issues) and [Discussions](https://github.com/open-mmlab/mmrotate/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmrotate.readthedocs.io/en/1.x/notes/4_faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmrotate) or [latest version (1.x)](https://github.com/open-mmlab/mmrotate/tree/dev-1.x).

### Task

I'm using the official example scripts/configs for the officially supported tasks/models/datasets.

### Branch

master branch https://github.com/open-mmlab/mmrotate

### Environment

```sys.platform: linux
Python: 3.8.16 (default, Jan 17 2023, 23:13:24) [GCC 11.2.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 3090
CUDA_HOME: /home/sysu/data/environment/openmmlab
NVCC: Cuda compilation tools, release 11.6, V11.6.124
GCC: gcc (Ubuntu 9.5.0-1ubuntu1~22.04) 9.5.0
PyTorch: 1.13.1
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.6
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.6, CUDNN_VERSION=8.3.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.14.1
OpenCV: 4.2.0
MMCV: 1.7.0
MMCV Compiler: GCC 9.3
MMCV CUDA Compiler: 11.6
MMRotate: 0.3.4+dc8b866
```
**Linux+pytorch1.13.1+cuda11.6+mmcv1.7.0**

### Reproduces the problem - code sample

```
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=2,
    train=dict(pipeline=train_pipeline, version=angle_version),
    val=dict(version=angle_version),
    test=dict(version=angle_version))
optimizer = dict(
    _delete_=True,
    type='AdamW',
    lr=0.0000125, #/0.0001 8*gpu_number,
    betas=(0.9, 0.999),
    weight_decay=0.05)
```
the whole file plz check it in the log uploaded below.

### Reproduces the problem - command or script

```
python tools/train.py \
     configs/lsknet/lsk_s_fpn_1x_dota_le90.py \
     --work-dir work_dirs/20240521_baseline_lsk_s_ss
```

### Reproduces the problem - error message

[20240521_143749.log](https://github.com/zcablii/LSKNet/files/15397745/20240521_143749.log)

### Additional information

single machine single GPU3090

1. reproduce1: lsk_s [split_ss_dota]

- batch_size=1
- lr=1.25e-05
- 12 epochs
- Expectation:  achieve map:0.78 on val-set. However, I only got 0.721

2. reproduce2: lsk_s [split_ms_dota]
- batch_size=1
- lr=1.25e-05
- 12 epochs
- Expectation:  achieve map:0.85 on val-set. However, I only got 0.616

I have reviewed other issues, mainly focusing on the adjustment of lr  and the number of GPUs, the gap is not particularly large, but I have a serious drop here.

**Comment by zcablii** [comment]
I believe the primary issue is your use of a batch size of 1. This can lead to unstable batch statistics during training. I recommend increasing your batch size and trying again.

**Comment by CrazyBrick** [comment]
> I believe the primary issue is your use of a batch size of 1. This can lead to unstable batch statistics during training. I recommend increasing your batch size (and adjusting your learning rate accordingly) and trying again.

but in your lsk_s_fpn_1x_dota_le90.py, you set the samples_per_gpu=1(I think this is the batchsize, is it right) , should I set bz=2 or larger?

**Comment by zcablii** [comment]
> > I believe the primary issue is your use of a batch size of 1. This can lead to unstable batch statistics during training. I recommend increasing your batch size (and adjusting your learning rate accordingly) and trying again.
> 
> but in your lsk_s_fpn_1x_dota_le90.py, you set the samples_per_gpu=1(I think this is the batchsize, is it right) , should I set bz=2 or larger?

But I use 8 GPUs with SyncBN, which means bs 8.

**Comment by CrazyBrick** [comment]
Did I misunderstand something? At least the result I found in your readme is: data=dict（
Samples_per_gpu=2,
Workers_per_gpu=2,

> > > I believe the primary issue is your use of a batch size of 1. This can lead to unstable batch statistics during training. I recommend increasing your batch size (and adjusting your learning rate accordingly) and trying again.
> > 
> > 
> > but in your lsk_s_fpn_1x_dota_le90.py, you set the samples_per_gpu=1(I think this is the batchsize, is it right) , should I set bz=2 or larger?
> 
> But I use 8 GPUs with SyncBN, which means bs 8.

Did I misunderstand something? At least the result I found in your readme is: 
```data=dict（
Samples_per_gpu=2,
Workers_per_gpu=2,
...
gpu_number = 8
optimizer = dict(
    type='AdamW', lr=2.5e-05, betas=(0.9, 0.999), weight_decay=0.05)


and you got map=0.810
[``your log of lsk_s-ms`](https://download.openmmlab.com/mmrotate/v1.0/lsknet/lsk_s_fpn_1x_dota_le90/lsk_s_fpn_1x_dota_le90_20230116.log)

**Comment by CrazyBrick** [comment]
> Our released checkpoint and model weights correspond to the released log and the results reported in our paper. However, we later empirically found that a better batch size and learning rate setting can easily achieve 81.8+ mAP. Therefore, we updated the configuration.



hi，thank you so much, I tried two experiments, and your suggestion was effective. Adding batch_size or lr can both improve the map (lsknet_s+bs=2 or lsknet_t+lr*2，achieve 0.77~0.78)

Also, I would like to ask about the use of mmrotate, as it seems that you have changed the backbone, which requires a lot of debugging work? But I think there are too many encapsulation in Mmrotate, not just simple model.fit or model.train. I would like to simplify some parts first to help me understand and then make further modifications. Do you think this is feasible?
I don't know where to go next, and it's a bit difficult for me to use. Do you have any suggestions or learning methods? (I raised an issue in the original repo (but it has not been maintained for a while) [issue1035](https://github.com/open-mmlab/mmrotate/issues/1035)

**Comment by zcablii** [comment]
You are right, encapsulation in MMRotate hides many detailed implementations, but if your modification is at a higher level, it should not be difficult. For a new backbone, I developed a backbone network under another PyTorch framework and made some adaptations to integrate it into the MMRotate framework. Simply add a file under mmrotate/models/backbones, register the model with @ROTATED_BACKBONES.register_module(), and add the model to __init__.py. You can follow or mimic the approach used in other files at the same level in MMRotate. In your case, to customize a loss function, work under mmrotate/models/losses. Reviewing how existing losses are implemented will make the process clear.

**Comment by CrazyBrick** [comment]
> You are right, encapsulation in MMRotate hides many detailed implementations, but if your modification is at a higher level, it should not be difficult. For a new backbone, I developed a backbone network under another PyTorch framework and made some adaptations to integrate it into the MMRotate framework. Simply add a file under mmrotate/models/backbones, register the model with @ROTATED_BACKBONES.register_module(), and add the model to **init**.py. You can follow or mimic the approach used in other files at the same level in MMRotate. In your case, to customize a loss function, work under mmrotate/models/losses. Reviewing how existing losses are implemented will make the process clear.

I see.Thank you for your prompt reply
In fact, I want to understand the input and output passed by different layers.It may not be considered a high level because I want to extract some middle  information. Or can you share the debugging code of lsknet before migrating to mmrotate (if available)

**Comment by zcablii** [comment]
One native and primary approach is to inject code to save intermediate features into local files, then perform analysis by loading these files. (This method is what I am using to visualize intermediate layer spatial attention.)

**Comment by CrazyBrick** [comment]
> One native and primary approach is to inject code to save intermediate features into local files, then perform analysis by loading these files. (This method is what I am using to visualize intermediate layer spatial attention.)


Good idea, I will give it a try!
